sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala

```diff
@@ -94,4 +108,21 @@ object InternalRow {
 
   /** Returns an empty [[InternalRow]]. */
   val empty = apply()
+
+  /**
+   * Copies the given value if it's string/struct/array/map type.
+   */
+  def copyValue(value: Any): Any = {
+    if (value.isInstanceOf[UTF8String]) {
+      value.asInstanceOf[UTF8String].copy()
+    } else if (value.isInstanceOf[InternalRow]) {
+      value.asInstanceOf[InternalRow].copy()
+    } else if (value.isInstanceOf[ArrayData]) {
+      value.asInstanceOf[ArrayData].copy()
+    } else if (value.isInstanceOf[MapData]) {
+      value.asInstanceOf[MapData].copy()
+    } else {
```

Comment:
Should we also support `Clonable`? This might be an escape door for developers who are putting their own objects in `InternalRow`?